services/web/server/src/simcore_service_webserver/meta_handlers.py

```diff
@@ -0,0 +1,276 @@
+        raise web.HTTPNotFound(
+            reason=f"No iterations found for project {_project_uuid=}/{commit_id=}"
+        )
+
+    assert len(selected_project_iterations) <= _limit  # nosec
+
+    # parse and validate response ----
+    iterations_items = [
+        ProjectIterationAsItem(
+            name=f"projects/{_project_uuid}/checkpoint/{commit_id}/iterations/{iter_id}",
+            parent=ParentMetaProjectRef(project_id=_project_uuid, ref_id=commit_id),
+            wcopy_project_id=wcp_id,
+            wcopy_project_url=url_for(
+                "get_project",
+                project_id=wcp_id,
+            ),
+            url=url_for(
+                f"{__name__}._list_meta_project_iterations_handler",
+                project_uuid=_project_uuid,
+                ref_id=commit_id,
+            ),
+        )
+        for wcp_id, iter_id in selected_project_iterations
+    ]
+
+    page = Page[ProjectIterationAsItem].parse_obj(
+        paginate_data(
+            chunk=iterations_items,
+            request_url=request.url,
+            total=total_number_of_iterations,
+            limit=_limit,
+            offset=_offset,
+        )
+    )
+    return web.Response(
+        text=page.json(**RESPONSE_MODEL_POLICY),
+        content_type="application/json",
+    )
+
+
+@routes.post(
+    f"/{VTAG}/projects/{{project_uuid}}/checkpoint/{{ref_id}}/iterations",
+    name=f"{__name__}._create_meta_project_iterations_handler",
+)
+@permission_required("project.snapshot.create")
+async def _create_meta_project_iterations_handler(request: web.Request) -> web.Response:
+    # FIXME: check access to non owned projects user_id = request[RQT_USERID_KEY]
+    url_for = create_url_for_function(request)
+    vc_repo = VersionControlForMetaModeling(request)
+
+    _project_uuid = ProjectID(request.match_info["project_uuid"])
+    _ref_id = request.match_info["ref_id"]
+    try:
+        commit_id = CommitID(_ref_id)
+    except ValueError as err:
+        # e.g. HEAD
+        raise NotImplementedError(
+            "cannot convert ref (e.g. HEAD) -> commit id"
+        ) from err
+
+    # core function ----
+    project_iterations = await create_or_get_project_iterations(
+        vc_repo, _project_uuid, commit_id
+    )
+
+    # parse and validate response ----
+    iterations_items = [
+        ProjectIterationAsItem(
+            name=f"projects/{_project_uuid}/checkpoint/{commit_id}/iterations/{iter_id}",
+            parent=ParentMetaProjectRef(project_id=_project_uuid, ref_id=commit_id),
+            wcopy_project_id=wcp_id,
+            wcopy_project_url=url_for(
+                "get_project",
+                project_id=wcp_id,
+            ),
+            url=url_for(
+                f"{__name__}._create_meta_project_iterations_handler",
+                project_uuid=_project_uuid,
+                ref_id=commit_id,
+            ),
+        )
+        for wcp_id, iter_id in project_iterations
+    ]
```

Comment:
your observation is correct. This is a poor implementation of pagination since it uses the entire list.

Iterations could potentially be infinite (e.g. a never ending loop) so pagination is critical here.

I guess I created this as a draft and was not modified. I will change it.

Reply:
It requires longer refactoring at the database level therefore, for the sake of clarity, I noted it in #2735